qiskit_nature/circuit/library/ansaetze/ucc.py

```diff
@@ -0,0 +1,173 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+TODO.
+"""
+
+from functools import partial
+from typing import Callable, List, Optional, Tuple, Union
+
+import logging
+
+from qiskit_nature.components.variational_forms.excitation_op_builder import ExcitationOpBuilder
+from .evolved_operator_ansatz import EvolvedOperatorAnsatz
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: Added to pass mypy, need change to real class
+class QubitOpConverter:
+    """ The QubitOpConverter """
+
+    def convert(self, op):
+        """TODO."""
+        pass
+
+
+class UCC(EvolvedOperatorAnsatz):
+    """The Unitary Coupled-Cluster Ansatz."""
+
+    EXCITATION_TYPE = {
+        's': 1,
+        'd': 2,
+        't': 3,
+        'q': 4,
+    }
+
+    # TODO: expose evolution and reps here, too?
+    def __init__(self, qubit_op_converter: QubitOpConverter,
+                 num_particles: Optional[Tuple[int, int]] = None,
+                 num_spin_orbitals: Optional[int] = None,
+                 excitations: Optional[Union[str, int, List[int], List[Callable]]] = None):
```

Comment:
We should be able to improve the Callable here to specify arguments and return types. Why would we want a list here - would a single callable not suffice? If a user wants to break down their code to have separate functions for single etc they can make one overall callable that calls the parts and builds the complete list.